text/0000-futures.md

```diff
@@ -0,0 +1,811 @@
+        Ok(())
+    }
+}
+
+pub struct Task { .. }
+
+// this impl is in `std` only:
+impl From<Box<dyn Future<Item = ()> + Send>> for Task { .. }
+
+/// Provides the reason that an executor was unable to spawn.
+pub struct SpawnError { .. }
+
+impl SpawnError {
+    /// Spawning is failing because the executor has been shut down.
+    pub fn shutdown() -> SpawnError;
+
+    /// Check whether this error is the `shutdown` error.
+    pub fn is_shutdown(&self) -> bool;
+
+    // additional error variants added over time...
+}
+```
+
+We need the executor trait to be usable as a trait object, which is why `Task`
+is constructed here from a boxed future. (In the no_std section, we'll see
+another constructor). In the long run, though, once we can take `dyn` by value,
+we would deprecate `BoxExecutor` and have:
+
+```rust
+trait Executor {
+    fn spawn(&mut self, task: Future<Item = ()> + Send) -> Result<(), SpawnError>;
+    fn status(&self) -> Result<(), SpawnError> { .. }
+}
+
+impl<E: BoxExecutor> Executor for E {
+    /* implement by boxing */
+}
+```
+
+This is why the RFC proposes the name `BoxExecutor` for the trait.
+
+In addition to the above, the `core::task` module will include the following API
+for helping detect bugs:
+
+```rust
+/// Marks the current thread as being within the dynamic extent of an
+/// executor.
+///
+/// Executor implementations should call this function before beginning to
+/// execute a tasks, and drop the returned `Enter` value after completing
+/// task execution:
+///
+/// ```rust
+/// let enter = enter().expect("...");
+/// /* run task */
+/// drop(enter);
+/// ```
+///
+/// Doing so ensures that executors aren't accidentally invoked in a nested fashion.
+///
+/// # Error
+///
+/// Returns an error if the current thread is already marked, in which case the
+/// caller should panic with a tailored error message.
+pub fn enter() -> Result<Enter, EnterError>
+```
+
+As stated in the doc comment, the expectation is that all executors will wrap
+their task execution within an `enter` to detect inadvertent nesting.
+
+### Task contexts
+
+All tasks are executed with two pieces of contextual information:
+
+- A `Waker` for waking the task up later on.
+- An executor, which is the "default" place to spawn further tasks.
+
+Notably, this list does *not* include task-local data; that can be addressed
+externally, as we'll see in a later section.
+
+The `core::task::Context` type gathers the (stack-rooted) contextual information
+together, and is passed by mutable reference to all polling functions:
+
+```rust
+/// Information about the currently-running task.
+///
+/// Contexts are always tied to the stack, since they are set up specifically
+/// when performing a single `poll` step on a task.
+pub struct Context<'a> { .. }
+
+impl<'a> Context<'a> {
+    /// Note: this signature is future-proofed for `E: Executor` later.
+    pub fn new<E>(waker: &'a Waker, executor: &'a mut E) -> Context<'a>
```

Comment:
I don't understand how this is the case, if this is stabilised with argument that's going to change, how is that future proof? is it because `Executor` will be a supertype/parent of `BoxExecutor`, thus the type is only being opened up to things it couldn't have been before?

Reply:
Along those lines, yes -- because there will be a blanket impl from `BoxExecutor` to `Executor`.

Reply:
thanks, I was confused on my first read-thru, but realising that it's an argument that will effectively be a subtype of the eventual trait.